File: client/src/main/java/su/plo/voice/client/audio/source/ClientDirectSource.java
package su.plo.voice.client.audio.source;

import gg.essential.universal.UMinecraft;
import net.minecraft.client.player.LocalPlayer;
import org.jetbrains.annotations.NotNull;
import su.plo.voice.api.client.PlasmoVoiceClient;
import su.plo.voice.api.client.audio.device.AlAudioDevice;
import su.plo.voice.api.client.audio.device.DeviceException;
import su.plo.voice.api.client.audio.device.source.AlSource;
import su.plo.voice.api.client.audio.device.source.DeviceSource;
import su.plo.voice.client.config.ClientConfig;
import su.plo.voice.proto.data.audio.source.DirectSourceInfo;

public final class ClientDirectSource extends BaseClientAudioSource<DirectSourceInfo> {

    public ClientDirectSource(@NotNull PlasmoVoiceClient voiceClient,
                              @NotNull ClientConfig config) {
        super(voiceClient, config);
    }

    @Override
    public void initialize(DirectSourceInfo sourceInfo) throws DeviceException {
        super.initialize(sourceInfo);
        updateSourceParams();
    }

    @Override
    protected float[] getPosition(float[] position) {
        if (sourceInfo.getRelativePosition() != null) {
            if (sourceInfo.isCameraRelative()) {
                position[0] = (float) sourceInfo.getRelativePosition().getX();
                position[1] = (float) sourceInfo.getRelativePosition().getY();
                position[2] = (float) sourceInfo.getRelativePosition().getZ();
            } else {
                return getAbsoluteSourcePosition(position);
            }
        } else {
            position[0] = 0F;
            position[1] = 0F;
            position[2] = 0F;
        }

        return position;
    }

    @Override
    protected float[] getLookAngle(float[] lookAngle) {
        // todo: lookAngle?
        lookAngle[0] = 0F;
        lookAngle[1] = 0F;
        lookAngle[2] = 0F;

        return lookAngle;
    }

    @Override
    protected boolean shouldCalculateOcclusion() {
        return false; // todo: relative position occlusion
    }

    private float[] getAbsoluteSourcePosition(float[] position) {
        if (sourceInfo.getRelativePosition() == null) return position;

        LocalPlayer player = UMinecraft.getPlayer();
        if (player == null) return position;

        position[0] = (float) (player.getX() + sourceInfo.getRelativePosition().getX());
        position[1] = (float) (player.getY() + player.getEyeHeight() + sourceInfo.getRelativePosition().getY());
        position[2] = (float) (player.getZ() + sourceInfo.getRelativePosition().getZ());

        return position;
    }

    private void updateSourceParams() {
        for (DeviceSource source : sourceGroup.getSources()) {
            if (source instanceof AlSource) {
                AlSource alSource = (AlSource) source;
                AlAudioDevice device = (AlAudioDevice) alSource.getDevice();

                device.runInContext(() -> {
                    alSource.setInt(
                            0x202, // AL_SOURCE_RELATIVE
                            sourceInfo.isCameraRelative() ? 1 : 0
                    );
                });
            }
        }
    }
}

File: client/src/main/kotlin/su/plo/voice/client/audio/source/ClientDirectSource.kt
package su.plo.voice.client.audio.source

import kotlinx.coroutines.runBlocking
import net.minecraft.client.Minecraft
import net.minecraft.client.player.LocalPlayer
import net.minecraft.world.phys.Vec3
import su.plo.config.entry.BooleanConfigEntry
import su.plo.config.entry.DoubleConfigEntry
import su.plo.voice.api.client.PlasmoVoiceClient
import su.plo.voice.api.client.audio.device.AlAudioDevice
import su.plo.voice.api.client.audio.device.DeviceException
import su.plo.voice.api.client.audio.device.source.AlSource
import su.plo.voice.client.config.VoiceClientConfig
import su.plo.voice.client.utils.toVec3
import su.plo.voice.proto.data.audio.source.DirectSourceInfo
import su.plo.voice.proto.packets.tcp.clientbound.SourceAudioEndPacket
import su.plo.voice.proto.packets.udp.clientbound.SourceAudioPacket

class ClientDirectSource(
    voiceClient: PlasmoVoiceClient,
    config: VoiceClientConfig,
    sourceInfo: DirectSourceInfo
) : BaseClientAudioSource<DirectSourceInfo>(voiceClient, config, sourceInfo) {

    override var sourceVolume: DoubleConfigEntry = createSourceVolume(sourceInfo)

    private var sourceMute: BooleanConfigEntry? = createSourceMute(sourceInfo)

    @Throws(DeviceException::class)
    override fun update(sourceInfo: DirectSourceInfo) {
        if (sourceInfo.sender != this.sourceInfo.sender) {
            sourceVolume = createSourceVolume(sourceInfo)
            sourceMute = createSourceMute(sourceInfo)
        }

        super.update(sourceInfo)
        runBlocking { updateSourceParams() }
    }

    override fun process(packet: SourceAudioPacket) {
        if (sourceMute?.value() == true) return
        super.process(packet)
    }

    override fun process(packet: SourceAudioEndPacket) {
        if (sourceMute?.value() == true) return
        super.process(packet)
    }

    override fun canHear(): Boolean {
        return isActivated()
    }

    override fun getPosition(): Vec3 {
        if (sourceInfo.relativePosition != null) {
            return if (sourceInfo.isCameraRelative) {
                sourceInfo.relativePosition!!.toVec3()
            } else {
                getAbsoluteSourcePosition()
            }
        }

        return Vec3.ZERO
    }

    override fun getLookAngle(): Vec3 {
        if (sourceInfo.lookAngle == null || sourceInfo.relativePosition == null)
            return Vec3.ZERO

        return sourceInfo.lookAngle!!.toVec3()
    }

    override fun shouldCalculateOcclusion() =
        sourceInfo.relativePosition != null && super.shouldCalculateOcclusion()

    override fun isPanningDisabled(): Boolean {
        return sourceInfo.isCameraRelative || super.isPanningDisabled()
    }

    override fun calculateDistanceGain(sourceDistance: Double, maxDistance: Double) =
        if (sourceInfo.isCameraRelative) 1.0
        else super.calculateDistanceGain(sourceDistance, maxDistance)

    private fun createSourceMute(sourceInfo: DirectSourceInfo): BooleanConfigEntry? =
        sourceInfo.sender?.let {
            config.voice
                .volumes
                .getMute("source_" + it.id)
        }

    private fun createSourceVolume(sourceInfo: DirectSourceInfo) =
        sourceInfo.sender?.let {
            config.voice
                .volumes
                .getVolume("source_${it.id}")
        } ?: super.sourceVolume

    private fun getAbsoluteSourcePosition(): Vec3 {
        return sourceInfo.relativePosition?.let {
            val player: LocalPlayer = Minecraft.getInstance().player ?: return Vec3.ZERO

            return Vec3(
                player.x + it.x,
                player.y + player.eyeHeight + it.y,
                player.z + it.z
            )
        } ?: Vec3.ZERO
    }

    private suspend fun updateSourceParams() {
        for (source in sourceGroup.sources) {
            if (source !is AlSource) continue

            val device = source.device as AlAudioDevice
            device.runInContext {
                source.setInt(
                    0x202,  // AL_SOURCE_RELATIVE
                    if (sourceInfo.isCameraRelative) 1 else 0
                )
            }
        }
    }
}
